src/LambdaBTree.h

```diff
@@ -0,0 +1,654 @@
+                        // a dynamic key doesn't make sense... tbh providence stuff should be
+                        // ignored in this version.
+                        return res;
+                    }
+
+                    // read result (atomic) -- just as a proof of concept, this is actually not valid!!
+                    std::atomic<typename Functor::result_type>& loc =
+                            *reinterpret_cast<std::atomic<typename Functor::result_type>*>(&(*pos).second);
+                    auto res = loc.load(std::memory_order_relaxed);
+
+                    // check validity
+                    if (!cur->lock.validate(cur_lease)) {
+                        // start over again
+                        return insert(k, hints, f);
+                    }
+
+                    // we found the element => return the result
+                    return res;
+                }
+
+                // get next pointer
+                auto next = cur->getChild(idx);
+
+                // get lease on next level
+                auto next_lease = next->lock.start_read();
+
+                // check whether there was a write
+                if (!cur->lock.end_read(cur_lease)) {
+                    // start over
+                    return insert(k, hints, f);
+                }
+
+                // go to next
+                cur = next;
+
+                // move on lease
+                cur_lease = next_lease;
+
+                continue;
+            }
+
+            // the rest is for leaf nodes
+            assert(!cur->inner);
+
+            // -- insert node in leaf node --
+
+            auto a = &(cur->keys[0]);
+            auto b = &(cur->keys[cur->numElements]);
+
+            auto pos = this->search.upper_bound(k, a, b, this->weak_comp);
+            auto idx = pos - a;
+
+            // early exit for sets
+            if (isSet && pos != a && this->weak_equal(*(pos - 1), k)) {
+                // validate result
+                if (!cur->lock.validate(cur_lease)) {
+                    // start over again
+                    return insert(k, hints, f);
+                }
+
+                // update provenance information
+                if (typeid(Comparator) != typeid(WeakComparator) && this->less(k, *(pos - 1))) {
+                    if (!cur->lock.try_upgrade_to_write(cur_lease)) {
+                        // start again
+                        return insert(k, hints, f);
+                    }
+                    this->update(*(pos - 1), k);
+
+                    // retrieve result before releasing lock
+                    auto res = (*(pos - 1)).second;
+
+                    cur->lock.end_write();
+                    // XXX: this provenance code should probably be removed, doesn't make
+                    // sense to run update on a dynamic key
+                    return res;
+                }
+
+                // read result (atomic) -- just as a proof of concept, this is actually not valid!!
```

Comment:
This comment is worrying. Why is this here if it is not valid?

Reply:
herbert's code, not sure actually - i returned it to the original retrieval of the value - the lock validation stuff *should* be fine to validate the read anyway.